README.md
# dMeasureCDM
Chronic Disease Management (CDM) module for dMeasure


NEWS.md
# News

# 0.6.0
11th July 2021

## New

* Aboriginal or Torres Strait Islander Health Assessment and age 45-49 year health assessments
  + `self$haatsi_list_cdm` - those who could qualify for health assessment
  + `self$ha4549_list_cdm` - those who could qualify for health assessment
  + suggestion of RN Cathy Xie, thanks
* Health Assessment for those with intellectual disability
  + `self$haiq_list_cdm` - those who could qualify for health assessment
* Refugee/Asylum seeker health assessment
  + `self$refugeeAsylum_list_cdm` - those who could qualify for health assessment
  
## Changes

* Aged 75+ Health Assessment renamed to `self$ha75_list_cdm`
  + display 'code' changed from 'HA' to 'HA75'
  
# 0.5.2
3rd July 2021

## New

* Filter `billings_cdm` by item status ('Never', 'Late', 'Up-to-date')
  + `item_status` list of status types defined as global in `CDM.R`
  + extra parameter `itemstatus_chosen` added to `billings_cdm`
  + suggestion of RN Dat Le, thanks
* `cdm_tem` and `cdm_item_names` defined as global in `CDM.R`

## Changes

* `dM$dateformat_choice` changes display format of dates

# 0.5.1
2nd December 2020

## Changes

* `steps_introduction_df` parameter `element_name` default to `as.character(NA)`
  + explicit `DailyMeasure::` for call to `steps_choose_contact_details`

# 0.5.0
25th July 2020

## New

* `dMeasureIntegration` to allow auto-loading

## Changes

* update `Suggests` in `DESCRIPTION`
* remove use of `shinycssloaders`

# 0.4.3
16th July 2020

## Changes

* `billings_cdm_list` requires (`shiny::req`) non-empty `dM$clinicians`
* change dropdown to dropMenu

# 0.4.2
8th July 2020

## Changes

* change dropdown label to 'Settings'

# 0.4.1
3rd July 2020

## Changes

* CDM items shown chosen through modal
  + as a result, the displayed lists of CDM items does not change each
    time a CDM item is selected/de-selected, but only after the modal
    is closed with 'OK'

# 0.4.0
27th June 2020

## Changes

* Responsiblity for GPstat/DailyMeasure UI moved from DailyMeasure
  to dMeasureCDM
  + in file `CDM_GPstatUI.R`
  + `shinydashboardmenuItem`, `dMeasureShinytabItems`, `datatableUI`, `datatableServer`
  
## Bugfix

* billings_cdm had an error if passed a non-null `intID` of length zero 
  e.g. numeric(0)
  
# 0.3.3
20th June 2020

* `billings_cdm` now gathers together Descriptions. If a CDM
  has been claimed before, the discovered reasons for doing a
  CDM are now included in the Description.

# 0.3.2
20th May 2020

* changes in management of lack of valid subscription/registration for
  chosen clinicians and selected date range.
  no longer tries to change the selected date range,
  instead returns empty dataframe.

# 0.3.1
10th April 2020

* add rintrojs introduction (steps_introduction_df)

# 0.3.0
30th March 2020

* contact view added to 'appointment' view
* add telehealth and telephone item numbers

# 0.2.0
2nd February 2020

* access `dMeasure$check_subscription` (subscription restrictions)


R/CDM_GPstatUI.R
##### Chronic Disease Management item numbers ###########################################
#' Chronic Disease Management (CDM) item numbers - UI for GPstat
#'
#' @name cdm_GPstatUI
#' @include CDM.R
#' R6 definitions
NULL

#' item description for left sidebar menu
#'
#' @name shinydashboardmenuItem
#'
#' @return shinydashboard menuItem object
#'
#' @export
shinydashboardmenuItem <- function() {
  return(
    shinydashboard::menuItem(
      "CDM items",
      tabName = "cdm", icon = shiny::icon("file-medical-alt")
    )
  )
}

#' center panel description
#'
#' @name dMeasureShinytabItems
#'
#' @return shinytabItems
#'
#' @export
dMeasureShinytabItems <- function() {
  x <- list(
    shinydashboard::tabItem(
      tabName = "cdm",
      shiny::fluidRow(column(
        width = 12, align = "center",
        h2("Chronic Disease Management items")
      )),
      shiny::fluidRow(column(
        width = 12,
        shiny::div(
          id = "cdm_datatable_wrapper",
          dMeasureCDM::datatableUI("cdm_dt")
        ))
      )
    )
  )
  return(x)
}

#' Chronic Disease Management (CDM) module - UI function
#'
#' Display CDM status and opportunities within selected range of
#' dates and providers
#'
#' @param id module ID (used in conjunction with 'callModule')
#'
#' @return Shiny user interface element
#'
#' @export
datatableUI <- function(id) {
  ns <- shiny::NS(id)

  shiny::tagList(
    shiny::fluidRow(
      shiny::column(
        4,
        shinyWidgets::switchInput(
          inputId = ns("printcopy_view"),
          label = paste(
            "<i class=\"fas fa-print\"></i>",
            "<i class=\"far fa-copy\"></i>",
            " Print and Copy View"
          ),
          labelWidth = "12em",
          width = "20em"
        )
      ),
      shiny::column(2,
                    offset = 2,
                    shinyWidgets::pickerInput(
                      inputId = ns("appointment_contact_view"),
                      choices = c("Appointment view", "Contact view"),
                      choicesOpt = list(icon = c("fa fa-calendar-alt", "fa fa-handshake"))
                    )
      ),
      shiny::column(2,
                    offset = 2, # note that total 'column' width = 12
                    shiny::uiOutput(ns("cdm_item_choice"))
      )
    ),
    shinycssloaders::withSpinner(
      DT::DTOutput(ns("cdm_table")),
      type = 8,
      hide.element.when.recalculating = FALSE,
      proxy.height = NULL
    )
  )
}

#' Chronic disease management list module - server
#'
#' chronic disease management items claimed, pending or unclaimed
#' for appointment list
#'
#' @param input as required by Shiny modules
#' @param output as required by Shiny modules
#' @param session as required by Shiny modules
#' @param dMCDM dMeasureCDM R6 object
#'  access to appointments lists, condition lists, history and EMR database
#'
#' @return none
#'
#' @export
datatableServer <- function(input, output, session, dMCDM) {
  ns <- session$ns

  # MBS (medicare benefits schedule) item numbers for CDM
  cdm_item <- data.frame(
    code = c(
      721, 92024, 92068, 723, 92025, 92069, 732, 92028, 92072,
      703, 705, 707,
      2517, 2521, 2525,
      2546, 2552, 2558,
      2700, 2701, 92112, 92124, 92113, 92125,
      2715, 2717, 92116, 92128, 92117, 92129
    ),
    name = c(
      "GPMP", "GPMP", "GPMP", "TCA", "TCA", "TCA", "GPMP R/V", "GPMP R/v", "GPMP R/V",
      "HA", "HA", "HA",
      "DiabetesSIP", "DiabetesSIP", "DiabetesSIP",
      "AsthmaSIP", "AsthmaSIP", "AsthmaSIP",
      "MHCP", "MHCP", "MHCP", "MHCP", "MHCP", "MHCP",
      "MHCP", "MHCP", "MHCP", "MHCP", "MHCP", "MHCP"
    )
  )

  cdm_item_names <- as.character(unique(cdm_item$name)) # de-factored

  output$cdm_item_choice <- renderUI({
    shinyWidgets::dropdown(
      input_id = "choice_dropdown",
      shinyWidgets::checkboxGroupButtons(
        inputId = ns("cdm_chosen"), label = "CDM items shown",
        choices = cdm_item_names,
        selected = setdiff(cdm_item_names, c("DiabetesSIP", "AsthmaSIP")),
        # DiabetesSIP and AsthmaSIP no longer valid items :(
        status = "primary",
        checkIcon = list(yes = icon("ok", lib = "glyphicon"))
      ),
      icon = icon("gear"),
      label = "CDM items shown"
    )
  })

  # filter to CDM item billed prior to (or on) the
  #  day of displayed appointments
  # only show most recent billed item in each category

  billings_cdm_list <-
    shiny::eventReactive(
      c(
        dMCDM$dM$appointments_filteredR(),
        dMCDM$dM$contact_count_listR(),
        input$cdm_chosen,
        input$appointment_contact_view,
        input$printcopy_view
      ), ignoreInit = TRUE, {
        shiny::validate(
          shiny::need(
            dMCDM$dMBillings$appointments_billingsR(),
            "No appointments defined"
          )
        )
        # respond to appointments_filteredR, since that is what is changed
        # when clinician or dates is changed
        if (input$appointment_contact_view == "Appointment view") {
          intID <- NULL
        } else {
          intID <- dMCDM$dM$contact_count_listR() %>>%
            dplyr::pull(InternalID)
        }

        billings_cdm_list <- dMCDM$billings_cdm(
          # called with 'appointment' method if intID = NULL,
          # otherwise called with 'intID' defined
          intID = intID, intID_Date = Sys.Date(),
          cdm_chosen = input$cdm_chosen,
          lazy = TRUE, # no need to re-calculate $appointments_billings
          screentag = !input$printcopy_view,
          screentag_print = input$printcopy_view
        )

        return(billings_cdm_list)
      }
    )

  ### create tag-styled datatable (or 'printable' datatable)

  cdm_styled_datatable <- shiny::reactive({
    shiny::validate(
      shiny::need(
        dMCDM$dM$appointments_filtered_timeR(),
        "No appointments in selected range"
      )
    )
    if (!is.null(billings_cdm_list()) &
        !is.null(dMCDM$dM$appointments_filtered_timeR())) {
      if (input$appointment_contact_view == "Appointment view") {
        cdm_list <- dMCDM$dM$appointments_filtered_timeR() %>>%
          dplyr::inner_join(billings_cdm_list(),
                            by = c(
                              "InternalID", "AppointmentDate",
                              "AppointmentTime", "Provider"
                            )
          )
        if (input$printcopy_view == TRUE) {
          # printable/copyable view
          DailyMeasure::datatable_styled(
            cdm_list %>>%
              dplyr::select(
                Patient, AppointmentDate, AppointmentTime,
                Provider, cdm_print
              ),
            colnames = c(
              "Patient", "Appointment Date", "Appointment Time",
              "Provider", "CDM items"
            )
          )
        } else {
          # fomantic/semantic tag view
          DailyMeasure::datatable_styled(
            cdm_list %>>%
              dplyr::select(
                Patient, AppointmentDate, AppointmentTime,
                Provider, cdm
              ),
            colnames = c(
              "Patient", "Appointment Date", "Appointment Time",
              "Provider", "CDM items"
            ),
            printButton = NULL, copyHtml5 = NULL,
            downloadButton = NULL, # no copy/print buttons
            escape = c(5)
          ) # only interpret HTML for last column
        }
      } else {
        # 'Contact view'
        intID <- billings_cdm_list() %>>%
          dplyr::pull(InternalID)
        cdm_list <- billings_cdm_list() %>>%
          dplyr::left_join(
            dMCDM$dM$db$patients %>>% # attach names etc.
              dplyr::filter(InternalID %in% c(intID, -1)) %>>%
              dplyr::select(
                InternalID, ExternalID, DOB,
                Firstname, Surname,
                HomePhone, MobilePhone
              ) %>>%
              dplyr::collect() %>>%
              dplyr::mutate(DOB = as.Date(DOB), Date = Sys.Date()) %>>%
              # initially Date is a dttm (POSIXt) object,
              # which makes the subsequent calc_age very slow,
              # and throws up warnings
              dplyr::mutate(
                Age = dMeasure::calc_age(DOB, Date),
                Patient = paste(Firstname, Surname)
              ),
            by = "InternalID"
          )
        if (input$printcopy_view == TRUE) {
          # printable/copyable view
          DailyMeasure::datatable_styled(
            cdm_list %>>%
              dplyr::select(
                Patient, ExternalID, DOB, Age,
                HomePhone, MobilePhone, cdm_print
              ),
            colnames = c(
              "Patient", "ExternalID", "DOB", "Age",
              "Home Phone", "Mobile Phone", "CDM items"
            )
          )
        } else {
          # fomantic/semantic tag view
          DailyMeasure::datatable_styled(
            cdm_list %>>%
              dplyr::select(
                Patient, ExternalID, DOB, Age,
                HomePhone, MobilePhone, cdm
              ),
            colnames = c(
              "Patient", "ExternalID", "DOB", "Age",
              "Home Phone", "Mobile Phone", "CDM items"
            ),
            printButton = NULL, copyHtml5 = NULL,
            downloadButton = NULL, # no copy/print buttons
            escape = c(7)
          ) # only interpret HTML for last column
        }
      }
    }
  })

  output$cdm_table <- DT::renderDT({
    cdm_styled_datatable()
  },
  server = TRUE
  )
}


R/introduction.R
# rintrojs introduction for CDM

#' rintrojs steps for CDM module
#'
#' @title steps_introduction_df
#' @param element_name the HTML datatable element, including datatable and helpers such as print/copy view
#' @description returns a dataframe of rintrojs steps
#'
#' requires pro-forma steps from DailyMeasure
#'
#' @export
steps_introduction_df <- function(element_name = as.character(NA)) {
  steps_df <-
    data.frame(
      element = as.character(NA),
      intro = c(paste(
        shiny::tags$h4("GPstat! Chronic Disease Management (CDM)"),
        shiny::br(),
        "View CDM opportunities for chosen clinicians and dates.",
        shiny::br(), shiny::br(),
        "CDM opportunities can be",
        "viewed for (usually future)",
        "Appointments", shiny::icon("calendar-alt"), ",",
        "or historical Contacts", shiny::icon("handshake"), ".",
        shiny::br(), shiny::br(),
        "If a selected clinician is not registered/subscribed",
        "then CDM opportunities will only be shown if at least ninety days old."
      )),
      position = "auto",
      stringsAsFactors = FALSE
    ) %>>%
    rbind(DailyMeasure::steps_choose_clinician_date_df()) %>>%
    rbind(data.frame(
      element = element_name,
      intro = c(paste(
        shiny::tags$h4("Chronic Disease Management (CDM) view"),
        shiny::br(),
        "List of CDM opportunities",
        "according to currently selected clinicians, dates",
        "and Appointment/Contact view.",
        shiny::br(), shiny::br(),
        "By default shows : ", shiny::strong("Patient"), "(name),",
        shiny::strong("DOB/Age"), " and ",
        shiny::strong("CDM items"), " list.",
        shiny::br(), shiny::br(),
        "In Appointments", shiny::icon("calendar-alt"), "view also",
        "shows Appointment details e.g.", shiny::strong("AppointmentTime"),
        "and", shiny::strong("Provider"), " (clinician).",
        shiny::br(), shiny::br(),
        "In Contacts", shiny::icon("handshake"), "view shows",
        shiny::strong("ExternalID"), "(Patient ID), and ",
        shiny::strong("telephone contact details")
      )),
      position = "auto"
    )) %>>%
    rbind(DailyMeasure::steps_datatable_helpers(element_name)) %>>%
    rbind(data.frame(
      element = element_name,
      intro = c(paste(
        shiny::tags$h4("CDM items shown"),
        shiny::br(),
        shiny::icon("gear"), shiny::br(),
        "Top-right of the table view.",
        shiny::br(), shiny::br(),
        "Choose CDM items to display.",
        shiny::br(), shiny::br(),
        "You can try it now",
        emo::ji("smile"), "!"
      )),
      position = "auto"
    )) %>>%
    rbind(data.frame(
      element = element_name,
      intro = c(paste(
        shiny::tags$h4("Appointment/Contact view"),
        shiny::br(),
        shiny::icon("calendar-alt"), shiny::icon("handshake"), shiny::br(),
        "Top/middle of the table view.",
        shiny::br(), shiny::br(),
        "Chronic Disease Management (CDM) opportunities can be",
        "viewed for (usually future)",
        "Appointments", shiny::icon("calendar-alt"), ",",
        "or historical Contacts", shiny::icon("handshake"), ".",
        shiny::br(), shiny::br(),
        "If using Contacts", shiny::icon("handshake"), "view",
        "then the", shiny::tags$em("types"), "of contact e.g.",
        "Appointment-book, Visits (progress notes) and/or Services (billings)",
        "and", shiny::tags$em("minimum number"), "of contact (within the",
        "specified time period) is chosen with the", shiny::tags$em("Contact details"),
        shiny::icon("handshake"), "tab in the right side-bar.",
        shiny::br(), shiny::br(),
        "You can try choosing between",
        shiny::icon("calendar-alt"), "Appointment view and",
        shiny::icon("handshake"), "Contact view now",
        emo::ji("smile"), "!"
      )),
      position = "auto"
    )) %>>%
    rbind(DailyMeasure::steps_choose_contact_details_df())

  return(steps_df)
}


R/userInterface.R
##### Chronic Disease Management item numbers ###########################################
#' Chronic Disease Management (CDM) item numbers - UI for GPstat
#'
#' @name cdm_GPstatUI
#' @include CDM.R
#' R6 definitions
NULL

#' item description for left sidebar menu
#'
#' @name shinydashboardmenuItem
#'
#' @return shinydashboard menuItem object
#'
#' @export
shinydashboardmenuItem <- function() {
  x <- list(
    shinydashboard::menuItem(
      "CDM items",
      tabName = "cdm",
      icon = shiny::icon("file-medical-alt")
    )
  )
  return(x)
}

#' center panel description
#'
#' @name dMeasureShinytabItems
#'
#' @return shinytabItems
#'
#' @export
dMeasureShinytabItems <- function() {
  x <- list(
    shinydashboard::tabItem(
      tabName = "cdm",
      shiny::fluidRow(column(
        width = 12, align = "center",
        h2("Chronic Disease Management items")
      )),
      shiny::fluidRow(column(
        width = 12,
        shiny::div(
          id = "cdm_datatable_wrapper",
          dMeasureCDM::datatableUI("CDM_dt")
        ))
      )
    )
  )
  return(x)
}

#' Chronic Disease Management (CDM) module - UI function
#'
#' Display CDM status and opportunities within selected range of
#' dates and providers
#'
#' @param id module ID (used in conjunction with 'callModule')
#'
#' @return Shiny user interface element
#'
#' @export
datatableUI <- function(id) {
  ns <- shiny::NS(id)

  shiny::tagList(
    shiny::fluidRow(
      shiny::column(
        width = 4,
        shinyWidgets::switchInput(
          inputId = ns("printcopy_view"),
          label = paste(
            "<i class=\"fas fa-print\"></i>",
            "<i class=\"far fa-copy\"></i>",
            " Print and Copy View"
          ),
          labelWidth = "12em",
          width = "20em"
        )
      ),
      shiny::column(
        width = 2,
        offset = 2,
        shinyWidgets::pickerInput(
          inputId = ns("appointment_contact_view"),
          choices = c("Appointment view", "Contact view"),
          choicesOpt = list(icon = c("fa fa-calendar-alt", "fa fa-handshake"))
        )
      ),
      shiny::column(
        width = 2,
        offset = 2, # note that total 'column' width = 12
        shiny::uiOutput(ns("cdm_item_choice"))
      )
    ),
    DT::DTOutput(ns("cdm_table"))
  )
}

#' Chronic disease management list module - server
#'
#' chronic disease management items claimed, pending or unclaimed
#' for appointment list
#'
#' @param id id
#' @param dMCDM dMeasureCDM R6 object
#'  access to appointments lists, condition lists, history and EMR database
#'
#' @return none
#'
#' @export
datatableServer <- function(id, dMCDM) {
  shiny::moduleServer(id, function(input, output, session) {
    ns <- session$ns

    cdm_chosen <- shiny::reactiveVal(
      setdiff(cdm_item_names, c("DiabetesSIP", "AsthmaSIP"))
    ) # cdm_item_names are defined in CDM.R as global
    itemstatus_chosen <- shiny::reactiveVal(
      unlist(item_status, use.names = FALSE)
      # item_status defined in CDM.R as global
    )
    output$cdm_item_choice <- renderUI({
      shinyWidgets::dropMenu(
        shiny::actionButton(
          inputId = ns("choice_dropdown"),
          icon = icon("gear"),
          label = "CDM settings"
        ),
        shiny::tags$div(
          shinyWidgets::checkboxGroupButtons(
            inputId = ns("cdm_chosen"), label = "CDM items shown",
            choices = cdm_item_names,
            selected = cdm_chosen(),
            # DiabetesSIP and AsthmaSIP no longer valid items :(
            status = "primary",
            checkIcon = list(yes = icon("ok", lib = "glyphicon"))
          ),
          shiny::hr(),
          shinyWidgets::checkboxGroupButtons(
            inputId = ns("itemstatus_chosen"), label = "Item status shown",
            choices = unlist(item_status, use.names = FALSE),
            # item_status defined in CDM.R
            selected = itemstatus_chosen(),
            status = "primary",
            checkIcon = list(yes = icon("ok", lib = "glyphicon"))
          ),
          shiny::br(),
          shiny::em("Close to confirm")
        ),
        placement = "bottom-end"
      )
    })
    shiny::observeEvent(
      input$choice_dropdown_dropmenu,
      ignoreInit = TRUE, {
        # this is triggered when shinyWidgets::dropMenu is opened/closed
        # tag is derived from the first tag in dropMenu, adding '_dropmenu'
        if (!input$choice_dropdown_dropmenu) {
          # only if closing the 'dropmenu' modal
          # unfortunately, is also triggered during Init (despite the ignoreInit)
          cdm_chosen(input$cdm_chosen)
          itemstatus_chosen(input$itemstatus_chosen)
        }
      }
    )

    # filter to CDM item billed prior to (or on) the
    #  day of displayed appointments
    # only show most recent billed item in each category

    billings_cdm_list <-
      shiny::eventReactive(
        c(
          dMCDM$dM$appointments_filteredR(),
          dMCDM$dM$contact_count_listR(),
          dMCDM$dM$dateformat(),
          cdm_chosen(),
          itemstatus_chosen(),
          input$appointment_contact_view,
          input$printcopy_view
        ), ignoreInit = FALSE, {
          shiny::req(dMCDM$dM$clinicians) # need some clinicians defined!
          shiny::validate(
            shiny::need(
              dMCDM$dMBillings$appointments_billingsR(),
              "No appointments defined"
            )
          )
          # respond to appointments_filteredR, since that is what is changed
          # when clinician or dates is changed
          if (input$appointment_contact_view == "Appointment view") {
            intID <- NULL
          } else {
            intID <- dMCDM$dM$contact_count_listR() %>>%
              dplyr::pull(InternalID)
          }

          billings_cdm_list <- dMCDM$billings_cdm(
            # called with 'appointment' method if intID = NULL,
            # otherwise called with 'intID' defined
            intID = intID, intID_Date = Sys.Date(),
            cdm_chosen = cdm_chosen(),
            itemstatus_chosen = itemstatus_chosen(),
            lazy = TRUE, # no need to re-calculate $appointments_billings
            screentag = !input$printcopy_view,
            screentag_print = input$printcopy_view
          )

          return(billings_cdm_list)
        }
      )

    ### create tag-styled datatable (or 'printable' datatable)

    cdm_styled_datatable <- shiny::reactive({
      shiny::validate(
        shiny::need(
          dMCDM$dM$appointments_filtered_timeR(),
          "No appointments in selected range"
        )
      )

      dateformat <- dMCDM$dM$formatdateR()
      # function to format date (is reactive)

      if (!is.null(billings_cdm_list()) &
          !is.null(dMCDM$dM$appointments_filtered_timeR())) {
        if (input$appointment_contact_view == "Appointment view") {
          cdm_list <- dMCDM$dM$appointments_filtered_timeR() %>>%
            dplyr::inner_join(
              billings_cdm_list(),
              by = c(
                "InternalID", "AppointmentDate",
                "AppointmentTime", "Provider"
              )
            )
          if (input$printcopy_view == TRUE) {
            # printable/copyable view
            DailyMeasure::datatable_styled(
              cdm_list %>>%
                dplyr::select(
                  Patient, AppointmentDate, AppointmentTime,
                  Provider, cdm_print
                ) %>>%
                dplyr::mutate(
                  AppointmentDate = dateformat(AppointmentDate)
                ),
              colnames = c(
                "Patient", "Appointment Date", "Appointment Time",
                "Provider", "CDM items"
              )
            )
          } else {
            # fomantic/semantic tag view
            DailyMeasure::datatable_styled(
              cdm_list %>>%
                dplyr::select(
                  Patient, AppointmentDate, AppointmentTime,
                  Provider, cdm
                ) %>>%
                dplyr::mutate(
                  AppointmentDate = dateformat(AppointmentDate)
                ),
              colnames = c(
                "Patient", "Appointment Date", "Appointment Time",
                "Provider", "CDM items"
              ),
              printButton = NULL, copyHtml5 = NULL,
              downloadButton = NULL, # no copy/print buttons
              escape = c(5)
            ) # only interpret HTML for last column
          }
        } else {
          # 'Contact view'
          intID <- billings_cdm_list() %>>%
            dplyr::pull(InternalID)
          cdm_list <- billings_cdm_list() %>>%
            dplyr::left_join(
              dMCDM$dM$db$patients %>>% # attach names etc.
                dplyr::filter(InternalID %in% c(intID, -1)) %>>%
                dplyr::select(
                  InternalID, ExternalID, DOB,
                  Firstname, Surname,
                  HomePhone, MobilePhone
                ) %>>%
                dplyr::collect() %>>%
                dplyr::mutate(DOB = as.Date(DOB), Date = Sys.Date()) %>>%
                # initially Date is a dttm (POSIXt) object,
                # which makes the subsequent calc_age very slow,
                # and throws up warnings
                dplyr::mutate(
                  Age = dMeasure::calc_age(DOB, Date),
                  Patient = paste(Firstname, Surname)
                ),
              by = "InternalID"
            )
          if (input$printcopy_view == TRUE) {
            # printable/copyable view
            DailyMeasure::datatable_styled(
              cdm_list %>>%
                dplyr::select(
                  Patient, ExternalID, DOB, Age,
                  HomePhone, MobilePhone, cdm_print
                ) %>>%
                dplyr::mutate(
                  DOB = dateformat(DOB)
                ),
              colnames = c(
                "Patient", "ExternalID", "DOB", "Age",
                "Home Phone", "Mobile Phone", "CDM items"
              )
            )
          } else {
            # fomantic/semantic tag view
            DailyMeasure::datatable_styled(
              cdm_list %>>%
                dplyr::select(
                  Patient, ExternalID, DOB, Age,
                  HomePhone, MobilePhone, cdm
                ) %>>%
                dplyr::mutate(
                  DOB = dateformat(DOB)
                ),
              colnames = c(
                "Patient", "ExternalID", "DOB", "Age",
                "Home Phone", "Mobile Phone", "CDM items"
              ),
              printButton = NULL, copyHtml5 = NULL,
              downloadButton = NULL, # no copy/print buttons
              escape = c(7)
            ) # only interpret HTML for last column
          }
        }
      }
    })

    output$cdm_table <- DT::renderDT({
      cdm_styled_datatable()
    },
    server = TRUE
    )
  })
}


R/CDM.R
#' dMeasureCDM - list billings module for dMeasure
#'
#' @md
#'
#' @name cdm
#' @title dMeasureCDM
#'
#' @include r6_helpers.R
#' functions to help create R6 classes
NULL

#' dMeasureIntegration
#'
#' @name dMeasureIntegration
#'
#' @description integration with dMeasure
#'   (especially DailyMeasure)
#'
#' @param information the information required
#'   `Provides` - modules provided (in this case, `dMeasureCDM`)
#'   `Requires` - the modules required (including `dMeasure`)
#'   `moduleID` - IDs of modules to create
#'
#' @return vector of required information
#'
#' @export
dMeasureIntegration <- function(information) {
  if (information == "Provides") {return(c("dMeasureCDM"))}
  if (information == "Requires") {return(c("dMeasure", "dMeasureBillings"))}
  if (information == "moduleID") {return(c("CDM_dt"))}
}

#' dMeasureCDM class
#' @title dMeasureCDM class
#' @description list chronic disease management opportunities
#' @export
dMeasureCDM <- R6::R6Class("dMeasureCDM",
  public = list(
    # dM is a dMeasure object
    dM = NULL,
    dMBillings = NULL,
    initialize = function(dMeasure_obj, dMeasureBillings_obj) {
      # dMeasure_obj is a R6 dMeasure object
      # dMeasureBilings_obj is a R6 object (which also depends on dMeasure!)
      self$dM <- dMeasure_obj
      self$dMBillings <- dMeasureBillings_obj
      if (length(public_init_fields$name) > 0) { # only if any defined
        for (i in 1:length(public_init_fields$name)) {
          if (public_init_fields$obj[[i]] == "dMeasureCDM") {
            self[[public_init_fields$name[[i]]]] <-
              eval(public_init_fields$value[[i]]) # could 'quote' the value
          }
        }
      }
      if (length(private_init_fields$name) > 0) { # only if any defined
        for (i in 1:length(private_init_fields$name)) {
          if (private_init_fields$obj[[i]] == "dMeasureCDM") {
            private[[private_init_fields$name[[i]]]] <-
              eval(private_init_fields$value[[i]]) # could 'quote' the value
          }
        }
      }

      if (requireNamespace("shiny", quietly = TRUE)) {
        # set reactive version only if shiny is available
        # note that this is for reading (from programs calling this object) only!
        if (length(reactive_fields$name) > 0) { # only if any .reactive() defined
          for (i in 1:length(reactive_fields$name)) {
            if (reactive_fields$obj[[i]] == "dMeasureCDM") {
              self[[reactive_fields$name[[i]]]] <- shiny::reactiveVal(
                eval(reactive_fields$value[[i]]) # could 'quote' the value
              )
            }
          }
        }
        if (length(reactive_event$name) > 0) { # only if any .reactive() defined
          for (i in 1:length(reactive_event$name)) {
            if (reactive_event$obj[[i]] == "dMeasureCDM") {
              self[[reactive_event$name[[i]]]] <-
                eval(reactive_event$value[[i]]) # could 'quote' the value
            }
          }
        }
      }
    }
  )
  # this is a 'skeleton' class
  # it is filled in the with the '.public' function
)

##### special reactive functions ##########################


.private(dMeasureCDM, "set_reactive", function(myreactive, value) {
  # reactive (if shiny/reactive environment is available) is set to 'value'
  # myreactive is passed by reference
  if (requireNamespace("shiny", quietly = TRUE)) {
    myreactive(value)
  }
})
.private(dMeasureCDM, "trigger", function(myreactive) {
  # toggles a reactive between (usually) 0 and 1
  if (requireNamespace("shiny", quietly = TRUE)) {
    myreactive(1 - shiny::isolate(myreactive()))
  }
})

##### constants #########################################################################

# MBS (medicare benefits schedule) item numbers for CDM
cdm_item <- data.frame(
  code = c(
    721, 92024, 92068, 723, 92025, 92069, 732, 92028, 92072,
    703, 705, 707,
    2517, 2521, 2525,
    2546, 2552, 2558,
    2700, 2701, 92112, 92124, 92113, 92125,
    2715, 2717, 92116, 92128, 92117, 92129,
    715, 228, 93470, 93479
  ),
  name = c(
    "GPMP", "GPMP", "GPMP", "TCA", "TCA", "TCA", "GPMP R/V", "GPMP R/v", "GPMP R/V",
    "HA", "HA", "HA",
    # note that the same item numbers 703/705/707 can be used for multiple health assessments
    # 75+, elsewhere coded as "HA75"
    # 45-49 (or 40-49) health assessment, which will elsewhere be code "HA45"
    # and *also* for intellectual disability health assessment, elsewhere coded "HAIQ"
    # and *also* for refugee health asseessment, elsewhere coded  'RHA'
    "DiabetesSIP", "DiabetesSIP", "DiabetesSIP",
    "AsthmaSIP", "AsthmaSIP", "AsthmaSIP",
    "MHCP", "MHCP", "MHCP", "MHCP", "MHCP", "MHCP",
    "MHCP", "MHCP", "MHCP", "MHCP", "MHCP", "MHCP",
    "HAATSI", "HAATSI", "HAATSI", "HAATSI"
  ),
  stringsAsFactors = FALSE
)

cdm_item_names <- c(as.character(unique(cdm_item$name)), "HA75", "HA45", "HAIQ", "RHA")
cdm_item_names <- cdm_item_names[!cdm_item_names == "HA"]
# de-factored and unique cdm_item$name
# need to manually add HA45 and HAIQ

item_status <- list(never = "Never", late = "Late", uptodate = "Up-to-date")

